stream: Readable.concat util method

tugrul opened the pull request:
There are many npm modules for this functionality and I think it is a common requirements because I see 452k weekly downloads of multistream npm package.

VoltrexMaster commented on doc/api/stream.md:
```diff
@@ -2077,6 +2077,27 @@ Calling `Readable.from(string)` or `Readable.from(buffer)` will not have
 
+### `stream.Readable.concat(streams, [options])`
+
+* `streams` {Readable[]|Iterable[]|AsyncIterable[]} List of readable streams or iterator of readable stream generator
```
```suggestion
* `streams` {Readable[]|Iterable[]|AsyncIterable[]} List of readable streams or
iterator of readable stream generator.
```
Every line must be 80 characters in length and this line should end with a period.

VoltrexMaster commented on doc/api/stream.md:
```diff
@@ -2077,6 +2077,27 @@ Calling `Readable.from(string)` or `Readable.from(buffer)` will not have
+
+* `streams` {Readable[]|Iterable[]|AsyncIterable[]} List of readable streams or iterator of readable stream generator
+* `options` {Object} Options provided to `new stream.Readable([options])`.
+  By default, `Readable.concat()` will set `options.objectMode` to `false`, unless
+  this is explicitly opted out by setting `options.objectMode` to `true`.
```
```suggestion
  By default, `Readable.concat()` will set `options.objectMode` to `false`,
  unless this is explicitly opted out by setting `options.objectMode` to `true`.
```
Every line must be 80 characters or less in length.

VoltrexMaster commented on doc/api/stream.md:
```diff
@@ -2077,6 +2077,27 @@ Calling `Readable.from(string)` or `Readable.from(buffer)` will not have
+  this is explicitly opted out by setting `options.objectMode` to `true`.
+* Returns: {stream.Readable}
+
+A utility method to concatenate multiple readable streams to single one. The main readable stream will consume the
+list of readable streams in order
```
```suggestion
A utility method to concatenate multiple readable streams to single one. The
main readable stream will consume the list of readable streams in order.
```
Every line must be 80 characters or less in length and this line must end with a period.

VoltrexMaster commented on lib/internal/streams/readable.js:
```diff
@@ -1371,6 +1371,18 @@ Readable.from = function(iterable, opts) {
+}
+
+Readable.concat = function(streams, opts) {
+  return Readable.from(streamSeriesIterator(streams), {objectMode: false, ...opts});
```
```suggestion
  return Readable.from(streamSeriesIterator(streams),
    { objectMode: false, ...opts });
```
Lint.

aduh95 commented on lib/internal/streams/readable.js:
```diff
@@ -1371,6 +1371,18 @@ Readable.from = function(iterable, opts) {
+  }
+}
+
+Readable.concat = function(streams, opts) {
```
To make `Readable.concat.length` equals to `1`:
```suggestion
Readable.concat = function(streams, opts = undefined) {
```

aduh95 commented on lib/internal/streams/readable.js:
```diff
@@ -1371,6 +1371,18 @@ Readable.from = function(iterable, opts) {
 
+async function * streamSeriesIterator(streams) {
+  for await (const stream of streams) {
+    for await (const chunk of stream) {
+      yield chunk;
+    }
```
Would that have the same effect?
```suggestion
    yield* steam;
```

tugrul replied:
This works fine and looks better. Thanks

tniessen commented:
cc @nodejs/streams

mcollina reviewed (changes requested):
Thanks for the PR!

Would you mind to add a few unit test for this feature?

I think if one of the "waiting" streams errors all should be destroyed, similar to how pipeline works.

ronag commented on lib/internal/streams/readable.js:
```diff
@@ -1371,6 +1371,17 @@ Readable.from = function(iterable, opts) {
+
+Readable.concat = function(streams, opts = undefined) {
+  return Readable.from(streamSeriesIterator(streams),
+                       { objectMode: false, ...opts });
```
IbjectMode should be true

ronag replied:
This doesn’t support old streams. See pipeline for how we fix that.

tugrul commented:
@mcollina I quickly lookup and only we need destroy remaining readable streams if it is an `Array` when error occurs on previous readable readings. This control not necessary for iterables because stream will initialize lazy. We do not have control of writable stream in this scenario. Following solutions can solve the problem.

```javascript
async function * streamSeriesIterator(streams) {
  let index = 0;

  try {
    for await (const stream of streams) {
      index++;
      yield* stream;
    }
  } finally {
    if (streams instanceof Array) {
      streams.slice(index).forEach(stream => stream.destroy());
    }
  }
}
```

or 

```javascript
async function * streamSeriesIterator(streams) {
  let index = 0;

  try {
    for await (const stream of streams) {
      index++;
      yield* stream;
    }
  } catch (err) {
    if (streams instanceof Array) {
      streams.slice(index).forEach(stream => stream.destroy(err));
    }
    
    throw err;
  }
}
```

jasnell commented:
As an alternative suggestion... consider adding this utility to the `streams/consumers` module and making it work with any `stream.Readable`, `ReadableStream`, and async iterable...

```js
import { concat, arrayBuffer } from 'node:stream/consumers';

const readable1 = new ReadableStream();
const readable2 = new ReadableStream();
const readable3 = new stream.Readable();

for await (const chunk of concat(readable1, readable2, readable3)) {
  console.log(chunk);
}

// Let's say you want all of the contents in a single ArrayBuffer...

await arrayBuffer(concat(readable1, readable2, readable3));
```

The return value would be just an async iterator. If someone wanted a `stream.Readable` from the results, it would be as simple as... `stream.Readable.from(concat(readable1, readable2, readable3))` ... getting a `ReadableStream` from it is a bit more complicated but also straightforward...

```js
const readable = new ReadableStream({
  async pull(c) {
    for await (const chunk of concat(readable1, readable2, readable3))
      c.enqueue(chunk);
    c.close();
  }
});
```

mcollina commented:
@tugrul unfortunately https://github.com/nodejs/node/pull/40075#issuecomment-917401863 does not work for node streams. They can error at any time even if they are not being consumed, you need to add on('error') handlers.

Ethan-Arrowood commented:
Hello, I'm doing some backlog hygiene on all PRs related to networking and streaming modules. There hasn't been any activity here for quite a while so I'm going to mark it `stalled`. It will auto-close in ~30 days. This is not a rejection; if you've like to pick it back up, rebase onto latest main and give me a ping.

Thank you!

tugrul committed:
stream: add concat()

Adds stream.concat(sources[, options]), which reads a sequence of
readable sources as a single async iterator. Each source is consumed
to completion before the next is started.

Returning an iterator rather than a Readable avoids having to pick an
objectMode on the caller's behalf: a concatenation of byte streams and
one of object streams want opposite defaults, and only the sources
know which case applies. Callers that want a stream can wrap the
result with Readable.from().

Sources are guarded as they are acquired rather than as they are read.
A Readable is live from construction, so a queued source that fails
while an earlier one is still draining would otherwise emit 'error'
with no listener attached and take down the process. concat() attaches
a handler to every source it holds a reference to, reports the first
failure, and destroys the rest.

Because a collection that already holds live streams and a producer
that creates one per iteration are indistinguishable, the input
protocol selects the behavior: a sync iterable is drained and guarded
up front, an async iterable is pulled one source at a time. Factories
stay inert until reached, and options.lazy overrides the default where
it guesses wrong.

Refs: https://github.com/nodejs/node/pull/40075
Assisted-By: Claude/Opus 5
Signed-off-by: Tuğrul Topuz <tugrultopuz@gmail.com>
